File: apps/backend/app/modules/production/journal_filters.py
"""Filtrer et trier le Journal — SERVEUR, sur toute l'histoire (addendum ADR-0034 « tri et filtre »).

## L'ordre est la décision : `WHERE` → `ORDER BY` → `LIMIT`

Filtrer une page déjà paginée est un défaut **silencieux**, et c'est la forme la plus coûteuse à
diagnostiquer : l'écran répondrait *« rien en maths »* alors que les lots de maths sont page 4.
Rien ici ne s'exécute côté Python sur des lots déjà chargés.

## Un filtre hiérarchique se résout en IDENTIFIANTS, puis devient un `IN`

`chapitre = C` doit retenir **aussi** les lots-pièce du chapitre C — un lot-pièce ne porte pas de
`chapter_id` (contrainte `ck_production_runs_exactly_one_scope`), il porte une notion. Sinon filtrer
par chapitre cacherait précisément les demandes de Massimo sur ce chapitre.

⚠️ **La règle « quelle est LA leçon de cette notion » n'est pas retraduite en SQL** : `lessons_by_skill`
(ADR-0037) est **appelée**, une fois par requête, et son résultat devient un paramètre. La récrire en
jointure referait le défaut qui a coûté un ADR entier — trois modules répondant différemment.

## La matière est une expression, pas deux requêtes

Deux chemins mènent à une matière (`chapitre → matière d'année` pour un lot-chapitre,
`Skill.subject_id` en direct pour un lot-pièce). Ils sont assemblés en **une** expression
`coalesce`, qui sert **à la fois** au filtre et au tri : sans ça, filtrer et trier pourraient ne pas
désigner la même chose.

⚠️ **La jointure matière passe par `school_year_subject_id`, exactement comme `lessons_by_skill`.**
Un chapitre rattaché seulement par `theme_id` n'a donc pas de matière ici — c'est **le même trou**
que dans le résolveur canonique, et le combler d'un seul côté créerait deux réponses. Voir la dette
nommée dans le rapport de slice : `Theme` porte une matière mais **aucune année scolaire**, donc la
réparation demande une décision de schéma, pas une jointure de plus.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, datetime, time, timedelta, timezone

from sqlalchemy import Select, and_, case, exists, func, or_, select
from sqlalchemy.orm import Session, aliased

from app.core.config import settings
from app.db.models import (
    Chapter,
    Lesson,
    LessonSkill,
    ProductionEvent,
    ProductionRun,
    SchoolYearSubject,
    Skill,
    Subject,
)

# Les cinq statuts que l'écran montre. La base n'en connaît que quatre : `stale` est RENDU
# (ADR-0034 §2), et il se filtre donc par un prédicat, jamais par une colonne.
STATUTS = ("queued", "running", "stale", "done", "failed")
MODES = ("manuel", "semi", "autonome", "sur_mesure", "inconnu")
TRIS = ("date", "matiere", "mode", "statut")


@dataclass(frozen=True)
class JournalFiltre:
    """Ce que Papa a demandé. Vide partout = aucun filtre, et c'est l'état d'ouverture."""

    subject_ids: tuple[int, ...] = ()
    chapter_ids: tuple[int, ...] = ()
    depuis: date | None = None
    jusqu_a: date | None = None
    statuts: tuple[str, ...] = ()
    modes: tuple[str, ...] = ()
    pieces: tuple[str, ...] = ()
    tri: str = "date"
    descendant: bool = True

    def actif(self) -> bool:
        return bool(
            self.subject_ids
            or self.chapter_ids
            or self.depuis
            or self.jusqu_a
            or self.statuts
            or self.modes
            or self.pieces
        )


@dataclass
class _Jointures:
    """Les alias partagés entre le filtre et le tri — une seule construction, une seule vérité."""

    chapitre: type
    matiere_annee: type
    notion: type
    sujet: type
    subject_id: object = field(init=False)

    def __post_init__(self) -> None:
        # Un lot porte SOIT un chapitre SOIT une notion. `coalesce` dit « celle des deux qui existe »
        # sans que l'appelant ait à connaître la contrainte.
        self.subject_id = func.coalesce(self.matiere_annee.subject_id, self.notion.subject_id)


def _joindre(stmt: Select, j: _Jointures) -> Select:
    """Les quatre jointures externes. **Externes** : un lot sans chapitre ni notion résolvable doit
    rester dans le flux non filtré — le faire disparaître serait réécrire l'histoire par omission."""
    return (
        stmt.outerjoin(j.chapitre, j.chapitre.id == ProductionRun.chapter_id)
        .outerjoin(j.matiere_annee, j.matiere_annee.id == j.chapitre.school_year_subject_id)
        .outerjoin(j.notion, j.notion.id == ProductionRun.scope_skill_id)
        .outerjoin(j.sujet, j.sujet.id == j.subject_id)
    )


def _alias() -> _Jointures:
    return _Jointures(
        chapitre=aliased(Chapter),
        matiere_annee=aliased(SchoolYearSubject),
        notion=aliased(Skill),
        sujet=aliased(Subject),
    )


def _notions_canoniques_du_chapitre(db: Session, chapter_ids: tuple[int, ...]) -> set[int]:
    """Les notions dont LA leçon (au sens ADR-0037) tombe dans l'un de ces chapitres.

    Deux temps, et ils ne sont pas interchangeables :

    1. les notions **candidates** — celles liées à une leçon de ces chapitres. Une simple jointure ;
    2. le **filtre canonique** — parmi elles, celles dont la *première* leçon rendue par
       `lessons_by_skill` est bien dans le chapitre. Une notion portée par deux leçons de chapitres
       différents ne doit compter que pour celui vers lequel Massimo est orienté.

    Sauter le temps 2 rendrait un sur-ensemble : le filtre ramènerait des lots-pièce dont la notion
    « appartient » ailleurs. Le résolveur est appelé **une fois**, sur les candidates seulement.
    """
    from app.modules.lesson_resolution import lessons_by_skill

    if not chapter_ids:
        return set()
    candidates = set(
        db.scalars(
            select(LessonSkill.skill_id)
            .join(Lesson, Lesson.id == LessonSkill.lesson_id)
            .where(Lesson.chapter_id.in_(chapter_ids))
        ).all()
    )
    if not candidates:
        return set()
    par_notion = lessons_by_skill(db, list(candidates))
    return {
        skill_id
        for skill_id, lecons in par_notion.items()
        if lecons and lecons[0].chapter_id in chapter_ids
    }


def _clause_statut(statuts: tuple[str, ...], *, maintenant: datetime):
    """`stale` = `running` dont le battement a expiré — et **`running` l'EXCLUT**.

    Sans cette exclusion, un lot zombie répondrait à deux filtres et Papa le compterait deux fois.
    L'affichage les sépare déjà (`run_status`) ; le filtre doit dire la même chose, ou l'un des deux
    ment.
    """
    limite = maintenant - timedelta(minutes=settings.production_heartbeat_timeout_minutes)
    perime = and_(
        ProductionRun.status == "running",
        ProductionRun.heartbeat_at.is_not(None),
        ProductionRun.heartbeat_at < limite,
    )
    clauses = []
    for statut in statuts:
        if statut == "stale":
            clauses.append(perime)
        elif statut == "running":
            clauses.append(and_(ProductionRun.status == "running", ~perime))
        else:
            clauses.append(ProductionRun.status == statut)
    return or_(*clauses)


def _couples_nommes() -> dict[str, tuple[int, int]]:
    """Les paliers de chaque régime, LUS dans `NIVEAUX` — jamais recopiés (ADR-0032/0037)."""
    from app.modules.settings.service import A0A, A1, NIVEAUX

    return {nom: (paliers[A0A], paliers[A1]) for nom, paliers in NIVEAUX.items()}


def _clause_mode(modes: tuple[str, ...]):
    """Le mode devient deux entiers — pur SQL, paginable, et stable dans le temps."""
    couples = _couples_nommes()
    inconnu = or_(ProductionRun.a0a_level.is_(None), ProductionRun.a1_level.is_(None))
    nomme = or_(
        *[
            and_(ProductionRun.a0a_level == a0a, ProductionRun.a1_level == a1)
            for a0a, a1 in couples.values()
        ]
    )
    clauses = []
    for mode in modes:
        if mode == "inconnu":
            clauses.append(inconnu)
        elif mode == "sur_mesure":
            # Des paliers renseignés qui ne composent AUCUN préréglage : un état légitime.
            clauses.append(and_(~inconnu, ~nomme))
        elif mode in couples:
            a0a, a1 = couples[mode]
            clauses.append(
                and_(ProductionRun.a0a_level == a0a, ProductionRun.a1_level == a1)
            )
    return or_(*clauses) if clauses else None


def _clause_piece(pieces: tuple[str, ...]):
    """Le type se lit dans les ÉVÉNEMENTS, pas dans les cinq tables de pièces.

    Une table au lieu de cinq, et elle porte **déjà** ses index (`run_id`,
    `ix_production_events_run_created`). Surtout : l'événement existe pour ce qui a été **produit**
    comme pour ce qui a été **sauté** ou a **échoué** — filtrer les tables de pièces n'aurait rendu
    que les succès, c'est-à-dire l'inverse de ce qu'on cherche dans un journal.

    ⚠️ **Un lot bloqué AVANT d'avoir touché une pièce ne répond à aucun filtre de type**, et c'est
    inévitable : `production_events.piece` est `NULL` sur `outcome='blocked'` (constat de code, pas
    un oubli). L'écran doit le DIRE dans son état vide.
    """
    return exists(
        select(ProductionEvent.id).where(
            ProductionEvent.run_id == ProductionRun.id,
            ProductionEvent.piece.in_(pieces),
        )
    )


def _rang_du_mode() -> tuple:
    """DEUX expressions, et la première est la décision.

    1. **hors de l'échelle** (`sur_mesure`, `inconnu`) : `0` ou `1`, trié **toujours croissant** ;
    2. le rang d'autonomie parmi les régimes nommés, qui suit le sens demandé.

    ⚠️ Un rang unique **ne suffit pas**. Avec un seul `case`, donner le rang maximal aux hors-échelle
    les met en fin en ordre croissant… et **en tête** en ordre décroissant — c'est-à-dire au sommet
    de l'autonomie, ce qu'ils ne sont pas. Ils ne sont ni plus ni moins autonomes que *Manual* : ils
    sont hors de l'échelle, et une échelle ne classe pas ce qui n'est pas dessus.

    (Trouvé en écrivant le test, pas après : la première version rendait `{sur_mesure, inconnu}` en
    tête du tri décroissant.)
    """
    couples = _couples_nommes()
    ordre = ["manuel", "semi", "autonome"]
    branches = [
        (
            and_(
                ProductionRun.a0a_level == couples[nom][0],
                ProductionRun.a1_level == couples[nom][1],
            ),
            rang,
        )
        for rang, nom in enumerate(ordre)
        if nom in couples
    ]
    nomme = case(*[(cond, 0) for cond, _ in branches], else_=1)
    return nomme, case(*branches, else_=len(ordre))


def appliquer(
    stmt: Select, db: Session, filtre: JournalFiltre, j: _Jointures, *, maintenant: datetime
) -> Select:
    """Le `WHERE`. Les critères se cumulent en ET ; plusieurs valeurs d'un critère, en OU."""
    stmt = _joindre(stmt, j)

    if filtre.subject_ids:
        stmt = stmt.where(j.subject_id.in_(filtre.subject_ids))

    if filtre.chapter_ids:
        notions = _notions_canoniques_du_chapitre(db, filtre.chapter_ids)
        chapitre_ou_piece = [ProductionRun.chapter_id.in_(filtre.chapter_ids)]
        if notions:
            chapitre_ou_piece.append(ProductionRun.scope_skill_id.in_(notions))
        stmt = stmt.where(or_(*chapitre_ou_piece))

    if filtre.depuis:
        stmt = stmt.where(
            ProductionRun.created_at >= datetime.combine(filtre.depuis, time.min, timezone.utc)
        )
    if filtre.jusqu_a:
        # Borne INCLUSE : « jusqu'au 4 août » doit retenir un lot de 15:52 le 4 août. Comparer à
        # minuit exclurait toute la journée demandée, et personne ne s'en apercevrait.
        stmt = stmt.where(
            ProductionRun.created_at <= datetime.combine(filtre.jusqu_a, time.max, timezone.utc)
        )

    if filtre.statuts:
        stmt = stmt.where(_clause_statut(filtre.statuts, maintenant=maintenant))
    if filtre.modes:
        clause = _clause_mode(filtre.modes)
        if clause is not None:
            stmt = stmt.where(clause)
    if filtre.pieces:
        stmt = stmt.where(_clause_piece(filtre.pieces))
    return stmt


def ordonner(stmt: Select, filtre: JournalFiltre, j: _Jointures) -> Select:
    """L'`ORDER BY`, et sa QUEUE — qui n'est pas une élégance.

    ⚠️ **Toute clé est départagée par `created_at DESC, id DESC`.** Sans cette queue, deux lots de
    même matière s'ordonnent différemment d'une page à l'autre, et la pagination **perd ou répète
    des lots** en silence. C'est un défaut de pagination classique, et son test existe.
    """
    if filtre.tri == "mode":
        # Le hors-échelle passe TOUJOURS en dernier, quel que soit le sens : il est trié à part.
        hors_echelle, rang = _rang_du_mode()
        principaux = [hors_echelle.asc(), rang.desc() if filtre.descendant else rang.asc()]
        return stmt.order_by(
            *principaux, ProductionRun.created_at.desc(), ProductionRun.id.desc()
        )

    cles = {
        "date": ProductionRun.created_at,
        "matiere": j.sujet.name,
        "statut": ProductionRun.status,
    }
    cle = cles.get(filtre.tri, ProductionRun.created_at)
    principal = cle.desc() if filtre.descendant else cle.asc()
    if filtre.tri == "date":
        return stmt.order_by(principal, ProductionRun.id.desc())
    # ⚠️ La queue reste TOUJOURS chronologique décroissante, même quand la clé est inversée : elle
    # départage, elle ne participe pas au sens du tri demandé.
    return stmt.order_by(principal, ProductionRun.created_at.desc(), ProductionRun.id.desc())


def compter(db: Session, filtre: JournalFiltre, *, maintenant: datetime) -> int:
    """Le total de l'ensemble FILTRÉ — jamais celui de l'histoire entière.

    `has_more` et le compteur d'écran portent tous deux sur cet ensemble : afficher « 7 sur 23 »
    quand le filtre en garde 7 est juste ; afficher « 7 sur 7 » cacherait qu'il y a autre chose.
    """
    j = _alias()
    stmt = appliquer(select(func.count(ProductionRun.id)), db, filtre, j, maintenant=maintenant)
    return db.scalar(stmt) or 0


def selectionner(db: Session, filtre: JournalFiltre, *, maintenant: datetime) -> Select:
    """La requête des lots : filtrée, triée, **et seulement ensuite** paginée par l'appelant."""
    j = _alias()
    stmt = appliquer(select(ProductionRun), db, filtre, j, maintenant=maintenant)
    return ordonner(stmt, filtre, j)

File: apps/frontend-papa/src/lib/journalFilters.ts
// Le filtre du Journal — son modèle, son vocabulaire d'écran, et son aller-retour avec l'URL.
//
// ⚠️ **Aucun libellé n'est inventé ici.** Les statuts viennent de `RUN_STATUS_LABEL`, les contenus
// de `PIECE_LABEL`/`PIECE_ICON`, les régimes de `NIVEAU_LABEL` (Manual · Hybrid · Autonom). Une
// seconde table de libellés divergerait de la première au premier renommage — c'est ce que
// l'addendum ADR-0032 §7.7 a coûté à réparer ailleurs.
//
// ⚠️ **Le filtrage est SERVEUR.** Rien de ce fichier ne filtre quoi que ce soit : il décrit ce que
// Papa a demandé, et le traduit en paramètres de requête. Filtrer les lots déjà chargés répondrait
// « rien en maths » alors que les lots de maths sont page 4.
import { type PieceKind } from "@zetis/types";

import { PIECE_ICON, PIECE_LABEL, RUN_STATUS_LABEL } from "./journal";
import { NIVEAU_LABEL } from "./settings";

/** Les cinq statuts que l'écran montre. `stale` est RENDU par le serveur, jamais stocké. */
export const STATUTS = ["queued", "running", "stale", "done", "failed"] as const;
export type StatutFiltre = (typeof STATUTS)[number];

/** Les régimes, plus les deux réponses qui ne sont pas des régimes. */
export const MODES = ["manuel", "semi", "autonome", "sur_mesure", "inconnu"] as const;
export type ModeFiltre = (typeof MODES)[number];

export const PIECES: PieceKind[] = ["cours", "fiche", "mindmap", "quiz", "srs"];

export const TRIS = ["date", "matiere", "mode", "statut"] as const;
export type TriCle = (typeof TRIS)[number];

export const TRI_LABEL: Record<TriCle, string> = {
  date: "Date",
  matiere: "Matière",
  mode: "Mode ZETIS",
  statut: "Statut",
};

/** ⚠️ Le vocabulaire d'ÉCRAN, jamais les clés serveur (`manuel | semi | autonome`). */
export const MODE_LABEL: Record<ModeFiltre, string> = {
  manuel: NIVEAU_LABEL.manuel,
  semi: NIVEAU_LABEL.semi,
  autonome: NIVEAU_LABEL.autonome,
  sur_mesure: "Sur mesure",
  inconnu: "Non enregistré",
};

export const STATUT_LABEL: Record<StatutFiltre, string> = {
  queued: RUN_STATUS_LABEL.queued!,
  running: RUN_STATUS_LABEL.running!,
  stale: RUN_STATUS_LABEL.stale!,
  done: RUN_STATUS_LABEL.done!,
  failed: RUN_STATUS_LABEL.failed!,
};

export const PIECE_FILTRE_LABEL = (kind: PieceKind) => `${PIECE_ICON[kind]} ${PIECE_LABEL[kind]}`;

export interface JournalFiltre {
  /** ⚠️ **MONO-matière**, et c'est un constat de code, pas une préférence : `SubjectFilterChips`
   *  est la brique partagée du Dashboard, de la Couverture et du Cahier de bord, et elle est
   *  contrôlée par `value: number | null`. La rendre multi toucherait trois autres pages — hors
   *  périmètre. Le serveur, lui, accepte déjà une liste : rien n'est perdu de ce côté. */
  subjectId: number | null;
  chapterId: number | null;
  depuis: string;
  jusquA: string;
  statuts: StatutFiltre[];
  modes: ModeFiltre[];
  pieces: PieceKind[];
  tri: TriCle;
  descendant: boolean;
}

export const FILTRE_VIDE: JournalFiltre = {
  subjectId: null,
  chapterId: null,
  depuis: "",
  jusquA: "",
  statuts: [],
  modes: [],
  pieces: [],
  tri: "date",
  descendant: true,
};

/** Combien de critères sont actifs — le compte que porte « Plus de filtres ».
 *
 * ⚠️ La matière n'y entre PAS : sa rangée reste visible, elle ne peut pas se cacher. Ce compte
 * répond à « qu'est-ce qui filtre sans que je le voie ? », pas à « combien de filtres ? ». */
export function criteresReplies(f: JournalFiltre): number {
  return (
    (f.chapterId !== null ? 1 : 0) +
    (f.depuis || f.jusquA ? 1 : 0) +
    (f.statuts.length > 0 ? 1 : 0) +
    (f.modes.length > 0 ? 1 : 0) +
    (f.pieces.length > 0 ? 1 : 0)
  );
}

export function filtreActif(f: JournalFiltre): boolean {
  return f.subjectId !== null || criteresReplies(f) > 0;
}

export function triParDefaut(f: JournalFiltre): boolean {
  return f.tri === "date" && f.descendant;
}

/** Bascule une valeur dans une liste — l'idiome des pilules multi-valeur. */
export function basculer<T>(liste: T[], valeur: T): T[] {
  return liste.includes(valeur) ? liste.filter((v) => v !== valeur) : [...liste, valeur];
}

// --- L'aller-retour avec l'URL ------------------------------------------------------------------
//
// ⚠️ Un journal filtré doit pouvoir être rouvert tel quel, et le retour arrière doit défaire le
// filtre plutôt que quitter la page. Mais **aucun filtre n'est actif à l'ouverture** : une page qui
// s'ouvrirait déjà filtrée cacherait son contenu à celui qui a oublié qu'il l'avait filtrée.

function lireListe<T extends string>(params: URLSearchParams, cle: string, permis: readonly T[]): T[] {
  return params.getAll(cle).filter((v): v is T => (permis as readonly string[]).includes(v));
}

export function depuisUrl(params: URLSearchParams): JournalFiltre {
  const nombre = (cle: string) => {
    const brut = params.get(cle);
    const valeur = brut === null ? Number.NaN : Number(brut);
    return Number.isFinite(valeur) ? valeur : null;
  };
  const tri = params.get("tri");
  return {
    subjectId: nombre("subject_id"),
    chapterId: nombre("chapter_id"),
    depuis: params.get("depuis") ?? "",
    jusquA: params.get("jusqu_a") ?? "",
    statuts: lireListe(params, "statut", STATUTS),
    modes: lireListe(params, "mode", MODES),
    pieces: lireListe(params, "piece", PIECES as readonly PieceKind[]),
    tri: (TRIS as readonly string[]).includes(tri ?? "") ? (tri as TriCle) : "date",
    descendant: params.get("sens") !== "asc",
  };
}

export function versUrl(f: JournalFiltre): URLSearchParams {
  const params = new URLSearchParams();
  if (f.subjectId !== null) params.set("subject_id", String(f.subjectId));
  if (f.chapterId !== null) params.set("chapter_id", String(f.chapterId));
  if (f.depuis) params.set("depuis", f.depuis);
  if (f.jusquA) params.set("jusqu_a", f.jusquA);
  f.statuts.forEach((s) => params.append("statut", s));
  f.modes.forEach((m) => params.append("mode", m));
  f.pieces.forEach((p) => params.append("piece", p));
  // Le tri par défaut ne s'écrit PAS dans l'URL : une URL propre est une URL sans filtre.
  if (f.tri !== "date") params.set("tri", f.tri);
  if (!f.descendant) params.set("sens", "asc");
  return params;
}
